crates/rpc/src/error.rs

```diff
@@ -168,45 +168,36 @@ impl ApplicationError {
             ApplicationError::CompiledClassHashMismatch => None,
             ApplicationError::UnsupportedTxVersion => None,
             ApplicationError::UnsupportedContractClassVersion => None,
-            ApplicationError::GatewayError(error) => Some(json!({
-                "error": error,
-            })),
+            ApplicationError::GatewayError(error) => Some(format!(
+                "gateway error {}: {}",
+                json!(error.code),
+                error.message,
+            )),
             ApplicationError::TransactionExecutionError {
                 transaction_index,
                 error,
-            } => Some(json!({
-                "transaction_index": transaction_index,
-                "execution_error": error,
-            })),
+            } => Some(format!(
+                "transaction execution error at transaction index {transaction_index}: {error}"
+            )),
             ApplicationError::Internal(_) => None,
             ApplicationError::Custom(cause) => {
                 let cause = cause.to_string();
                 if cause.is_empty() {
                     None
                 } else {
-                    Some(json!({
-                        "error": cause.to_string(),
-                    }))
+                    Some(cause.to_string())
                 }
             }
-            ApplicationError::NoTraceAvailable(error) => Some(json!({
-                "error": error,
-            })),
-            ApplicationError::ContractErrorV05 { revert_error } => Some(json!({
-                "revert_error": revert_error
-            })),
-            ApplicationError::TooManyKeysInFilter { limit, requested } => Some(json!({
-                "limit": limit,
-                "requested": requested,
-            })),
-            ApplicationError::UnexpectedError { data } => Some(json!({
-                "error": data,
-            })),
-            ApplicationError::ProofLimitExceeded { limit, requested } => Some(json!({
-                "limit": limit,
-                "requested": requested,
-            })),
-            ApplicationError::ValidationFailureV06(error) => Some(json!(error)),
+            ApplicationError::NoTraceAvailable(error) => Some(json!(error).to_string()),
+            ApplicationError::ContractErrorV05 { revert_error } => Some(revert_error.clone()),
+            ApplicationError::TooManyKeysInFilter { limit, requested } => Some(format!(
+                "too many keys in filter: limit is {limit}, requested {requested}"
+            )),
+            ApplicationError::UnexpectedError { data } => Some(data.clone()),
+            ApplicationError::ProofLimitExceeded { limit, requested } => Some(format!(
+                "proof limit exceeded: limit is {limit}, requested {requested}"
+            )),
+            ApplicationError::ValidationFailureV06(error) => Some(error.clone()),
```

Comment:
I think we should change _only_ `UnexpectedError` and `ValidationFailureV06` as those are the errors in the spec that have `data` typed as `string`: https://github.com/starkware-libs/starknet-specs/blob/49665932a97f8fdef7ac5869755d2858c5e3a687/api/starknet_write_api.json#L292 and https://github.com/starkware-libs/starknet-specs/blob/49665932a97f8fdef7ac5869755d2858c5e3a687/api/starknet_write_api.json#L259

`TransactionExecutionError` and `ContractErrorV05` are _definitely_ an object: https://github.com/starkware-libs/starknet-specs/blob/49665932a97f8fdef7ac5869755d2858c5e3a687/api/starknet_api_openrpc.json#L3903

Reply:
I think it's reasonable to keep things as they are for now, as technically all these changes are breaking our JSON-RPC API. For the cases where it does not match the spec it can actually be considered a fix -- however, I think we should probably leave all other cases alone (and maybe do the change once we introduce a new JSON-RPC version).